File: Python3/average_of_levels_in_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        nodes = [root]
        result = []
        
        while True:
            row_sum, row_count = 0, 0
            new_nodes = []
            
            for node in nodes:
                if not node: continue
                row_sum += node.val
                row_count += 1
                new_nodes.append(node.left)
                new_nodes.append(node.right)
                
            if row_count == 0: break
            result.append(row_sum / float(row_count))
            nodes = new_nodes
        
        return result

File: C/average_of_levels_in_binary_tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int height(struct TreeNode* root){
    if(root) return 1 + fmax(height(root->left),height(root->right));
    else return 0;
}

double BFS(struct TreeNode* root, int level, int* num_elem){
    if(!root) return 0;
    
    if(level == 0){
       *num_elem += 1;
        return root->val;
    } 
    else return BFS(root->left, level-1, num_elem) + BFS(root->right, level-1, num_elem);
}

double* averageOfLevels(struct TreeNode* root, int* returnSize){
    *returnSize = height(root);
    double* result = malloc(*returnSize * sizeof(double));
    int* numElem = malloc(*returnSize * sizeof(int));
    
    for(int i = 0; i < *returnSize; i++){
        numElem[i] = 0;
        result[i] = BFS(root, i, &numElem[i]);
        
        if(numElem[i] != 0) result[i] /= numElem[i];        
    }
    
    return result;
}
